docs/src/interface-definition.md

```diff
@@ -0,0 +1,227 @@
+
+Outputs for the functions need to have following properties
+
+- Energy is a subtype of `Number` that has a unit with dimensions of energy (mass * length^2 / time^2)
+- Force output is a subtype of `AbstractVector` with element type also a subtype of AbstractVector (length 3 in 3D) and unit with dimensions of force (mass * length / time^2). With additional property that it can be reinterpret as a matrix
+- Virial is a square matrix (3x3 in 3D) that has units of force times length or energy
+
+
+## Implementing the interface
+
+You can either implement both of the calls e.g. for energy
+
+`AtomsCalculators.potential_energy(system, calculator; kwargs...)` and
+`AtomsCalculators(AtomsCalculators.Energy(), system, calculator; kwargs...)`
+
+### Example implementations
+
+Example `potential_energy` implementation
+
+```julia
+using AtomsCalculators
+using Unitful
+struct MyType
+end
+
+AtomsCalculators.@generate_interface function AtomsCalculators.potential_energy(system, calculator::Main.MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    return 0.0u"eV"
+end
+```
+
+!!! note "Type definition under @generate_complement macro"
+    You need to use explicit definition of type when using
+    `@generate_interface` macro. `Main.MyType` is fine `MyType` is not!
+
+    You also need to define the type before the macro call.
+
+Completely equivalent implementation is
+
+```julia
+AtomsCalculators.@generate_interface function AtomsCalculators.calculate(::AtomsCalculators.Energy, system, calculator::Main.MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    return 0.0u"eV"
+end
+```
+
+Example `virial` implementation
+
+```julia
+AtomsCalculators.@generate_interface function AtomsCalculators.virial(system, calculator::Main.MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    return zeros(3,3) * u"eV"
+end
+```
+
+Equivalent implementation is
+
+```julia
+AtomsCalculators.@generate_interface function AtomsCalculators.calculate(::AtomsCalculators.Virial, system, calculator::Main.MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    return zeros(3,3) * u"eV"
+end
+```
+
+### Implementing forces call
+
+Basic example
+
+```julia
+AtomsCalculators.@generate_interface function AtomsCalculators.forces(system, calculator::Main.MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition
+    return zeros(AtomsCalculators.promote_force_type(system, calculator), length(system))
```

Comment:
Actually, why not have a function `zero_forces` instead that not only determines the element type, but also the array type (maybe useful later for GPU things or if there is sparsity that can be exploited or similar).

Reply:
This is a good idea. I added `zero_forces` and also kept `promote_force_type`.